harness/determined/pytorch/deepspeed/_deepspeed_trial.py

```diff
@@ -0,0 +1,881 @@
+            if self.env.test_mode:
+                break
+
+        all_keys = self.context.distributed._zmq_gather(keys if keys is None else list(keys))
+        if self.is_chief:
+            all_keys = [k for k in all_keys if k is not None]
+            keys = all_keys[0]
+        keys = self.context.distributed._zmq_broadcast(keys)
+
+        for callback in self.callbacks.values():
+            callback.on_validation_epoch_end(batch_metrics)
+
+        metrics = _metric_utils._reduce_metrics(
+            self.context.distributed,
+            batch_metrics=batch_metrics,
+            keys=keys,
+            metrics_reducers=_metric_utils._prepare_metrics_reducers(
+                pytorch.Reducer.AVG, keys=keys
+            ),
+        )
+        metrics.update(
+            _metric_utils._convert_metrics_to_numpy(self.context.reduce_metrics(for_training=False))
+        )
+
+        if self.context.distributed.size > 1 and any(
+            map(
+                lambda c: util.is_overridden(c.on_validation_end, pytorch.PyTorchCallback)
+                or util.is_overridden(c.on_validation_step_end, pytorch.PyTorchCallback),
+                self.callbacks.values(),
+            )
+        ):
+            logging.debug(
+                "Broadcasting metrics to all worker processes to execute a "
+                "validation step end callback"
+            )
+            metrics = self.context.distributed._zmq_broadcast(metrics)
+
+        for callback in self.callbacks.values():
+            if util.is_overridden(callback.on_validation_step_end, pytorch.PyTorchCallback):
+                logging.warning(
+                    "on_validation_step_end is now deprecated, please use on_validation_end instead"
+                )
+                callback.on_validation_step_end(metrics)
+
+        for callback in self.callbacks.values():
+            callback.on_validation_end(metrics)
+
+        if not self.is_chief:
+            return {}
+
+        num_inputs *= self.context.mpu.get_data_parallel_world_size()
+        logging.debug(
+            f"Done validating: {num_inputs} records in {self.num_validation_batches} batches."
+        )
+        return {"num_inputs": num_inputs, "validation_metrics": metrics}
+
+    def _load(self, load_path: pathlib.Path) -> None:
+        # Right now we will load all checkpoint shards on each node regardless of which
+        # checkpoints are needed.
+        # TODO (Liam): revisit later to optimize sharded checkpoint loading.
+
+        # Load stateful things tracked by Determined on all slots.
+        checkpoint: Optional[Dict[str, Any]] = None
+        ckpt_path = "det_state_dict.pth"
+        maybe_ckpt = load_path.joinpath(ckpt_path)
+        if maybe_ckpt.exists():
+            checkpoint = torch.load(str(maybe_ckpt), map_location="cpu")  # type: ignore
+        if checkpoint is None or not isinstance(checkpoint, dict):
```

Comment:
I see you copied this from pytorch trial, but if we load a checkpoint and it's not a dict, please raise an error.

Also, pytorch trial checks multiple paths, but you only need to check one.  `_load()` shouldn't be called if there's not a path to actually load.